### PR TITLE
Fix 'use explicit type' with ref variables.

### LINKED ISSUE #42880: The refactoring 'Use explicit type' should conserve the ref-ness of the variable.
**Version Used**: 
netcoreapp3.1 
VS 16.5.1

**Steps to Reproduce**:

```
static void Main()
{
     string str = "hello";

     ref var rStr1 = ref str; // Codefix wants to replace 'ref var' with 'string'

     ref readonly var rStr2 = ref str; // Codefix wants to replace 'ref readonly var' with 'string'

     Console.WriteLine($"{rStr1} {rStr2}");
}
```

**Expected Behavior**:
The codefix should conserve the ref-ness of the variable.
In this case, 'ref var' should be replaced with 'ref string'.

**Actual Behavior**:
The codefix discards the ref-ness of the variable, producing unexpected and invalid code.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes #42880

### CODE REVIEW

**Comment by ryzngard** [comment]
@y87feng want to take a look?

File: Roslyn.sln [hunk 1]
```diff
@@ -418,6 +418,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 	ProjectSection(SolutionItems) = preProject
 		.editorconfig = .editorconfig
 		.vsconfig = .vsconfig
+		NuGet.config = NuGet.config
```
Comment (CyrusNajmabadi) [inline Roslyn.sln]:
```suggestion
```

File: src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs [hunk 2]
```diff
@@ -48,19 +50,21 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             foreach (var diagnostic in diagnostics)
             {
-                var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
-                ReplaceTypeWithVar(editor, node);
+                var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true) as TypeSyntax;
+                if (node != null)
```
Comment (genlu) [inline src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs:54]: When do we hit the null path? Is there a test for this?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs [hunk 3]
```diff
@@ -339,9 +339,11 @@ private bool IsSafeToSwitchToVarWithoutNeedingSpeculation(DeclarationExpressionS
         }
 
         internal static ExpressionSyntax GetInitializerExpression(ExpressionSyntax initializer)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs:341]: > Looks like this is also used in codestyle layer, any test for this?

Yes :) all our features that switch to/from 'var' (including my new tests) :)

File: src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs:54]: no.  you're right.  this isn't needed. t hs has to be safe given teh construction of the diagnostic

File: src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs [hunk 4]
```diff
@@ -48,19 +50,21 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             foreach (var diagnostic in diagnostics)
             {
-                var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
-                ReplaceTypeWithVar(editor, node);
+                var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true) as TypeSyntax;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs:53]:
```suggestion
                var typeSyntax = (TypeSyntax)root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
```

File: src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs [hunk 5]
```diff
@@ -48,19 +50,21 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             foreach (var diagnostic in diagnostics)
             {
-                var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
-                ReplaceTypeWithVar(editor, node);
+                var typeSyntax = (TypeSyntax)root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
+                if (node != null)
+                    ReplaceTypeWithVar(editor, node);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseImplicitOrExplicitType/UseImplicitTypeCodeFixProvider.cs:55]:
```suggestion
                ReplaceTypeWithVar(editor, typeSyntax);
```

**Pull request merged**